extensions/oplogPopulator/allocationStrategy/ImmutableConnector.js

```diff
@@ -0,0 +1,57 @@
+const AllocationStrategy = require('./AllocationStrategy');
+
+/**
+ * @class ImmutableConnector
+ *
+ * @classdesc ImmutableConnector is an allocation
+ * strategy that always requires a new connector to be
+ * created for each bucket.
+ */
+class ImmutableConnector extends AllocationStrategy {
+    /**
+     * Assign the bucket to a new connector. If a
+     * connector has retained the bucket, use it.
+     * If the bucket is already assigned to a connector,
+     * return the connector.
+     * @param {String} bucket bucket name
+     * @returns {Promise<Connector>} connector
+     */
+    async getOrAssignBucketConnector(bucket) {
+        let connector;
+        if (this._bucketsToConnectors.has(bucket)) {
+            return this._bucketsToConnectors.get(bucket);
+        } else if (this._retainedBuckets.has(bucket)) {
+            connector = this._retainedBuckets.get(bucket);
+        } else if (!connector) {
+            connector = await this._connectorsManager.addConnector();
+        }
+        this._bucketsToConnectors.set(bucket, connector);
+        await connector.addBucket(bucket);
+        return connector;
+    }
+
+    /**
+     * Delete a bucket from a connector, if it can be
+     * deleted
+     * @param {string} bucket bucket name
+     * @returns {Promise<Connector | null>} the modified connector returned if modified
+     */
+    async removeBucketFromConnector(bucket) {
+        // If the pipelines are immutable, we only remove the bucket
+        // from the connector if it is the single bucket in use. In
+        // this case, the connector will be either deleted by the
+        // connectorManager, or reused, if a bucket with the same
+        // name is created. Otherwise, the connector would be
+        // reconfigured during the next connectorsManager update.
+        const connector = this._bucketsToConnectors.get(bucket);
+        if (connector?.buckets.length === 1) {
```

Comment:
if we have an "old" connector with multiple buckets, it will never be removed with this implementation?

not sure this is a _practical_ issue, but it leaves a unexpected behavior, possibly hitting us in the back later, when we least expect it...

Reply:
thinking on this, it seems we do not have the right "hooks" in the interface for handling removal:
* the `Retain` strategy needs to _know_ when a bucket is removed from the connector, just like it needs to know when the connectors configs are actually "applied"; but the other strategies do not need it --> this should just be another event (or call) from `ConnectorsManager`, only in this specific strategy
* what is part of the strategy is dedicing when to actually "change" the connector configuration, i.e. if the actual update should be skipped (`connector.updatePipeline()`) : so maybe we should simply have a `canUpdate(connector) -> bool` method instead, which gets called from whereever appropriate (conenctorsManager maybe?)

NOTE:
* Indeed, it makes the `AllocationStrategy` not deal only with allocation, but also with the actual applying of configuration... but should be a single strategy to ensure consistency, I think... This however means the strategy should probably be a parameter of the Allocator, and the "choice" of strategy moved to oplogPopulator
* The "extra retained buckets" metric may thus be nicely generated in this new `canUpdate()` metric: as the function will naturally see when it prevents actually 'removing' buckets from a connector, and we will thus listen to more buckets than needed...